### DISCUSSION TITLE
Websocket Questions

### DISCUSSION THREAD

**Discussion by tsmith123** [ACINQ/phoenixd#39] [General]
Could someone please clarify a couple of things for me...

Firstly, is authentication required to connect to the websocket endpoint? If not, how come? If it is then what's the approach?

Secondly, is the websocket connection used for notification of invoices _paid_ (as an alternative to webhooks) or just invoices _created_? I'm not quite sure reading from the docs.

**Comment by pm47** [comment]
> Firstly, is authentication required to connect to the websocket endpoint? 

Yes, with basic auth

> If it is then what's the approach?

Here is an example with `websocat`, a command line utility for websocket:
```
$ # no auth
$ websocat ws://127.0.0.1:9740/websocket
websocat: WebSocketError: WebSocketError: Received unexpected status code (401 Unauthorized)
websocat: error running

$ # correct auth
$ websocat --basic-auth :d106c828ff86c4e2dbfa565470fec41a5f67334c834535a411e9b7289e42473a ws://127.0.0.1:9740/websocket
{     "type": "payment_received",     "amountSat": 12345,     "paymentHash": "505c830d825e5f6dcbb0c7bffdce254ff94d3da3eb7463c28b0aa1760df412a6",     "externalId": null }
{     "type": "payment_received",     "amountSat": 6789,     "paymentHash": "8c0fa93d1870546faed3f08ac923ba3690a55bd678773c2bc2cb53ccae0fb054",     "externalId": null }
```

> is the websocket connection used for notification of invoices paid (as an alternative to webhooks) or just invoices created?

It's for payments received.

**Reply by tsmith123** [reply]
> Here is an example with websocat, a command line utility for websocket:

Thanks. So what's actually happening with the auth flow as I'm not familiar with websocat. Basic auth with standard ws clients isn't really a "thing" as you can't set the headers. Typically, to authenticate against a websocket server you would attach some sort of token (probably retrieved from a protected endpoint) on the end of the connection url and an upgrade will occur.

> It's for payments received.

So it's an alternative to the webhook flow then right?

**Reply by pm47** [reply]
> Typically, to authenticate against a websocket server you would attach some sort of token (probably retrieved from a protected endpoint) on the end of the connection url and an upgrade will occur.

What client are you using? I'm not a websocket expert, but AFAIK authentication is pushed to the HTTP layer and can use any standard technique, see the rfc: https://datatracker.ietf.org/doc/html/rfc6455.html#section-10.5.

> So it's an alternative to the webhook flow then right?

Exactly, use whichever you prefer, they are redundant. Websocket is maybe more exotic in the backend, but it makes the communication "one way" between your backend and phoenixd and saves you from setting up an endpoint.

**Reply by tsmith123** [reply]
So I would typically use something like this: https://github.com/websockets/ws

As websocket and basic auth don't go together, websocat must be doing something with that `--basic-auth` flag that's passed. Perhaps it's attaching it to the end of the connection string so it becomes `ws://127.0.0.1:9740/websocket?auth=12345` for example. I guess I just need to know how phoenixd expects the auth value string to be passed and I will do that from my client.

**Reply by tsmith123** [reply]
I've managed to connect to the websocket endpoint now 👍

**Reply by remyers** [reply]
I ran into this problem trying to create a basic web page using web sockets. As @tsmith123 mentioned, you can't set headers in javascript and there does not seem to be a solution in sight based on this SO post:  [https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api](https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api).

Would it be possible to pass authorization in the url instead? or perhaps use the ugly hack "Smuggle access tokens inside Sec-WebSocket-Protocol" mentioned in the SO post?

**Comment by tsmith123** [comment]
Assuming you're trying to connect from a Node app you can use something like this:

```js
const socket = new WebSocket(url, {
  headers: {
    Authorization:
      "Basic " + Buffer.from(":" + password).toString("base64"),
    },
});
```

This exmple uses `ws` npm package.

**Comment by remyers** [comment]
I did something similar using the WebSocket package for Python:

```python
import asyncio
import signal
import websockets

async def shim(websocket):
    uri = "ws://token:a6a3c41245dec52d57af36dd2887dce0334b7e85381a15696455cf7f0734e5d2@127.0.0.1:9740/websocket"
    async with websockets.connect(uri) as websocket1:
        # Close the connection when receiving SIGTERM.
        loop = asyncio.get_running_loop()
        loop.add_signal_handler(
            signal.SIGTERM, loop.create_task, websocket1.close())

        # Process messages received on the connection.
        async for message in websocket1:
            print(f"Received: {message}")
            await websocket.send(message)

async def main():
    async with websockets.serve(shim, "localhost", 8765):
        await asyncio.Future()  # run forever

if __name__ == "__main__":
    asyncio.run(main())
```
I'm trying to create something like a ticker that shows received payments. This shim should let me create a local browser page that connects to the web socket created by my python shim at port 8765.
